### PR TITLE
Article on What is rest api 

### PR DESCRIPTION
**Pull request by D-stevpm**
#359 

- [ ] I have checked that my suggestion follows the Contribution guide

- [ ]  I have added which issue this PR closes

- [ ] I have 0 warnings from Danger.js

- [ ] If I have added images, they are in /files/ folder

### CODE REVIEW

File: practices/what_is_rest_api [hunk 1]
```diff
@@ -0,0 +1,79 @@
+---
+slug: what_is_rest_api
```
Comment (prokopsimek) [inline practices/what_is_rest_api:2]:
```suggestion
slug: rest-api
```

File: practices/what_is_rest_api [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+stages:
+  - research
```
Comment (prokopsimek) [inline practices/what_is_rest_api:4]:
```suggestion
  - development
```

File: practices/what_is_rest_api [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  - restful api
+  - http methods
+  - api endpoints
+  - json
+  - api design
+  - api documentation
+author_github_username: D-stevpm
+---
+
+# What is RESTful API
```
Comment (prokopsimek) [inline practices/what_is_rest_api:24]:
```suggestion
# REST API
```

File: practices/what_is_rest_api [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+![RESTful API](/files/what_is_rest_api.png)
+
+The majority of applications in use today are created for web and mobile platforms. Application Programming Interfaces (APIs) for the web have consequently gained importance. Among the different types of web APIs, REST (Representational State Transfer) API has become the most popular due to its simplicity, scalability, and flexibility.
+
+##What is an API
```
Comment (prokopsimek) [inline practices/what_is_rest_api:30]:
```suggestion
### What is an API 
```

File: practices/what_is_rest_api [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  - api design
+  - api documentation
+author_github_username: D-stevpm
+---
+
+# What is RESTful API
+
+![RESTful API](/files/what_is_rest_api.png)
+
+The majority of applications in use today are created for web and mobile platforms. Application Programming Interfaces (APIs) for the web have consequently gained importance. Among the different types of web APIs, REST (Representational State Transfer) API has become the most popular due to its simplicity, scalability, and flexibility.
```
Comment (prokopsimek) [inline practices/what_is_rest_api:28]:
```suggestion
**TL;DR**

The majority of applications in use today are created for web and mobile platforms. Application Programming Interfaces (APIs) for the web have consequently gained importance. Among the different types of web APIs, REST (Representational State Transfer) API has become the most popular due to its simplicity, scalability, and flexibility.
```

File: practices/what_is_rest_api [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+There have been a lot of definitions of API from industry leaders, after going through these numerous definitions I can simply describe APIs as a state of mechanisms that enable two software components to communicate with each other using a set of definitions and protocols. 
+
+According to ***Mastering API Architecture book by James Gough, Daniel Bryant & Matthew Auburn*** describe API as: 
+
+*  An API represents an abstraction of the underlying implementation.
+*  An API is represented by a specification that introduces types. Developers can understand the specifications and use tooling to generate code in multiple languages to implement an API consumer (software that consumes an API).
+*  An API has defined semantics or behavior to model the exchange of information effectively.
+*  Effective API design enables the extension to customers or third parties for business integration.
+
+##What is REST API?
```
Comment (prokopsimek) [inline practices/what_is_rest_api:41]:
```suggestion
## What is the REST API?
```

File: practices/what_is_rest_api [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+A RESTful API is based on representational state transfer (REST), an architectural style and approach to communications that is frequently used in the building of web services. It is also known as a RESTful web service or REST API. RESTful APIs are also referred to as ***web services***.
+
+Safe online information exchange between two computer systems is made possible by the RESTful API. To carry out various duties, the majority of business apps must interface with other internal and external applications. For instance, in order to automate invoicing and connect with an internal timesheet application, your internal accounts system must exchange information with your customers' banking system in order to generate monthly payslips. This information transmission is supported by RESTful APIs because they adhere to standards for safe, dependable, and effective software communication
+
+###How to Implement RESTful API
```
Comment (prokopsimek) [inline practices/what_is_rest_api:49]: Please add a section "Why You Might Want the REST API" and "Problems the REST API Helps to Solve" before this section. 

According to https://github.com/DXHeroes/knowledge-base-content/blob/master/CONTRIBUTING.md#content-composition

File: practices/what_is_rest_api [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+*  **Hypermedia as the engine of application state (HATEOAS)**: This principle states that the client should be able to navigate the API using hyperlinks included in the response. This allows the API to evolve without breaking existing clients.
+
+**Advantages of REST API**
+
+*  **Scalability**: RESTful APIs are highly scalable due to their stateless nature, which allows for easy scaling of resources.
+*  **Flexibility**: RESTful APIs are flexible in terms of the format of data, which allows for easy integration with other web-based systems.
+*  **Easy to understand**: RESTful APIs are easy to understand due to their simplicity and uniform interface.
+*  **Easy to develop**: RESTful APIs are easy to develop due to their uniform interface and stateless nature, which reduces the complexity of the application.
+*  **Platform independence**: RESTful APIs can be used with any programming language and platform, making them highly interoperable.
+
```
Comment (prokopsimek) [inline practices/what_is_rest_api:79]: Please, add missing sections "Common Pitfalls of the REST API" and "Resources for the REST API" according to https://github.com/DXHeroes/knowledge-base-content/blob/master/CONTRIBUTING.md#content-composition

File: files/how_restful_api_works.png
Comment (prokopsimek) [inline files/how_restful_api_works.png:1]: This image describes URLs in the form of `/allUsers`, `/newUser` and `/updateUser` but the best practice for REST API is to use `GET /users`, `POST /users` and `PUT /users`. Can you please change the image for example this (https://www.sitepoint.com/creating-restful-apis-express-4/#creating-the-api-for-the-movie-app) that describes the format of URL better? Thank you

**Review by prokopsimek — CHANGES_REQUESTED** [review]
@D-stevpm Hi Steve, thank you for your article. The content is amazing! I am proposing to update some sections and please some sections according to the content composition are missing. Are you able to add those sections, please? Let me know if you need some help. :)

**Comment by D-stevpm** [comment]
I will work on it and definitely let you know when I need assistance

**Comment by D-stevpm** [comment]
Hi @prokopsimek I have updated the following section
- Why You Might Want the REST API
- Problems the REST API Helps to Solve
- Common Pitfalls of the REST API
- Resources for the REST API
- Also I changed the image as suggested, KIndly review it and let me know if there is something to adjust

File: practices/what_is_rest_api (hunk 8 above)
Comment (D-stevpm) [reply practices/what_is_rest_api:112]: Completed

**Comment by prokopsimek** [comment]
@D-stevpm Hi, the article is amazing now! Is there any reason why you closed it? :) Thanks!

**Comment by D-stevpm** [comment]
@prokopsimek Please let me know what you think about the new #396
